modules/dbparser/radix.c

```diff
@@ -131,22 +131,52 @@ typedef struct _RParserPCREState
 {
   pcre *re;
   pcre_extra *extra;
+  gint *matches;
+  gsize matches_size;
 } RParserPCREState;
 
 gboolean
 r_parser_pcre(gchar *str, gint *len, const gchar *param, gpointer state, RParserMatch *match)
 {
   RParserPCREState *self = (RParserPCREState *) state;
   gint rc;
-  gint matches[2];
 
-  rc = pcre_exec(self->re, self->extra, str, strlen(str), 0, 0, matches, 2);
-  if (rc <= 0)
-    return FALSE;
-  *len = matches[1] - matches[0];
+  rc = pcre_exec(self->re, self->extra, str, strlen(str), 0, 0, self->matches, self->matches_size);
+
+  if (rc == PCRE_ERROR_NOMATCH)
+    {
+      return FALSE;
+    }
+
+  if (rc < 0)
+    {
+      msg_error("Error while matching regexp", evt_tag_int("error_code", rc));
+      return FALSE;
+    }
+
+  if (rc == 0)
+    {
+      msg_error("Error while storing matching substrings");
+      return FALSE;
+    }
+
+  *len = self->matches[1] - self->matches[0];
   return TRUE;
 }
 
+static void
+_alloc_matches(RParserPCREState *self)
+{
+  gint num_matches;
+  if (pcre_fullinfo(self->re, self->extra, PCRE_INFO_CAPTURECOUNT, &num_matches) < 0)
+    g_assert_not_reached();
+  if (num_matches > RE_MAX_MATCHES)
+    num_matches = RE_MAX_MATCHES;
+
+  self->matches_size = 3 * (num_matches + 1);
+  self->matches = g_malloc(self->matches_size * sizeof(gint));
```

Comment:
Me seems similar to https://github.com/balabit/syslog-ng/pull/2827#discussion_r303752604. Can `matches` be part of `RParserPCREState`?

Reply:
The radix evaluation runs (or should be running in parallel), IIRC there was a lock just because of reload but eventually I would like to get rid off that. Querying the number of matches of a compiled regexp should be a very simple operation, here's the code from pcre_fullinfo():

```
  case PCRE_INFO_CAPTURECOUNT:
  *((int *)where) = re->top_bracket;
  break;
```

So apart from the function call overhead, this is basically nothing. Querying only the count should be fine, but the matches array should be allocated on the spot (hopefully with alloca instead of malloc), as the first one is much faster.

Reply:
Thanks, I moved the allocation to be done per message, and it is done with `alloca`.